--- MozillaBaseElement.rb.orig	2007-04-19 23:17:50.000000000 -0500
+++ MozillaBaseElement.rb	2007-04-19 23:19:54.000000000 -0500
@@ -1360,10 +1360,10 @@
                 # We get method name with trailing '=' when we try to assign a value to a 
                 # property. So just remove the '=' to get the type 
                 temp = ""
-                assingning_value = false
+                assigning_value = false
                 if(methodName =~ /(.*)=$/)
                     temp  = "#{element_object}.#{$1}" 
-                    assingning_value = true
+                    assigning_value = true
                 else
                     temp = "#{element_object}.#{methodName}"
                 end    
@@ -1373,8 +1373,10 @@
                 method_type = read_socket()
                 #puts "method_type is : #{method_type}"
 
-                if(assingning_value)
+                if(assigning_value)
                     if(method_type != "boolean")
+                        args[0].gsub!("\\", "\\"*4)
+                        args[0].gsub!("\"", "\\\"")
                         args[0].gsub!("\n","\\n")
                         jssh_command = "#{element_object}.#{methodName}\"#{args[0]}\""
                     else
